crates/wasmtime/src/component/linker.rs

```diff
@@ -230,6 +231,59 @@ impl<T> LinkerInstance<'_, T> {
         self.insert(name, Definition::Func(func.into_host_func()))
     }
 
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// `name` must refer to a function type import in `component`.  If and when
+    /// that import is invoked by the component, the specified `func` will be
+    /// called, which must return a `Val` which is an instance of the result
+    /// type of the import.
+    pub fn func_new<F: Fn(&[Val]) -> Result<Val> + Send + Sync + 'static>(
+        &mut self,
+        component: &Component,
+        name: &str,
+        func: F,
+    ) -> Result<()> {
+        self.func_new_with_context(component, name, move |_, args| func(args))
+    }
+
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// This is the same as [`LinkerInstance::func_new`] except that it accepts
+    /// a function which accepts a `StoreContextMut` as the first parameter.
+    pub fn func_new_with_context<
```

Comment:
Given that `func_new` is somewhat rare and already fairly verbose I think it's ok to skip the variant-without-`StoreContextMut` and go straight to requiring the argument during `func_new`